compiler/rustc_codegen_llvm/src/context.rs

```diff
@@ -385,11 +385,16 @@ impl MiscMethods<'tcx> for CodegenCx<'ll, 'tcx> {
                 } else {
                     "rust_eh_personality"
                 };
-                let fty = self.type_variadic_func(&[], self.type_i32());
-                self.declare_cfn(name, llvm::UnnamedAddr::Global, fty)
+                if let Some(llfn) = self.get_declared_value(name) {
```

Comment:
For the codegen unit that ends up being allocated the definition of the `rust_eh_personality` this will return the definition in the source code. For the other codegen units AFAICT the `else` branch will be executed the first time `personality_fn` is requested and then this will continue returning the default declaration.

I wonder if we might want to force private definition of the personality fn in every CU of every crate? That sounds like it'd be the only way to make sure we accurately specify the signature of the personality fn everywhere.

Reply:
I wonder if this might pose problems for e.g. LTO which gets insight into the mismatching signatures eventually...

Reply:
> I wonder if we might want to force private definition of the personality fn in every CU of every crate?

That is impossible. There are two different possible eh_personality functions that can be used. Either the one in libpanic_unwind or the one in libpanic_abort. This means that codegen isn't possible when using private eh_personality functions until you chose the final panic strategy. In addition both depend on at least libcore, so which crate do you compile first? Libcore has to be compiled before either as they depend on libcore, but libcore also has to be compiled after either as it needs eh_personality due to many functions needing to run cleanup on unwinding to drop locals.

Reply:
> OTOH we have lived with an arguably wrong "default" signature here for a very long time with no apparent problems. So it might be fine either way?

Wrong signatures can give problems with LTO and PGO AFAIK. In any case it is UB with LTO due to needing to merge two incompatible function signatures.

Reply:
There are no calls using incorrect signature, so I wouldn't expect this to be an issue. 

I also reviewed LLVM IR linking implementation and experimented a bit with intentionally mismatched signatures. As far as I can see that should be fine. Clang uses the same signature by the way.